advanced_functionality/causal-inference/container/causal_nex/predictor.py

```diff
@@ -0,0 +1,140 @@
+# This is the file that implements a flask server to do inferences. It's the file that you will modify to
+# implement the scoring for your own algorithm.
+
+from __future__ import print_function
+
+import io
+import json
+import ast
+import os
+import pickle
+import signal
+import sys
+import traceback
+from collections import defaultdict
+
+import flask
+import pandas as pd
+from causalnex.structure.structuremodel import StructureModel
+from causalnex.network import BayesianNetwork
+from causalnex.inference import InferenceEngine
+
+prefix = "/opt/ml/"
+model_path = os.path.join(prefix, "model")
+
+# A singleton for holding the model. This simply loads the model and holds it.
+# It has a predict and an intervention functions that does a prediction (or intervention) based on the model and the input data.
+
+
+class ScoringService(object):
+    model = None  # Where we keep the model when it's loaded
+
+    @classmethod
+    def get_model(cls, model_name="causal_model.pkl"):
+        """Get the model object for this instance, loading it if it's not already loaded."""
+        if cls.model == None:
+            with open(os.path.join(model_path, model_name), "rb") as inp:
+                cls.model = pickle.load(inp)
+        return cls.model
+
+    @classmethod
+    # pass node as param
+    def predict(cls, input, target_node):
+        """For the input, do the predictions and return them.
+
+        Args:
+            input (a list of dictionaries): The data on which to do the predictions. There will be
+                one prediction per row in the result dataframe"""
+        bn = cls.get_model()
+        result = bn.predict(
+            pd.DataFrame.from_dict(input, orient="columns"), target_node
+        )
+
+        return result
+
+    @classmethod
+    def intervention(cls, input):
+
+        """Users can apply an intervention to any node in the data, updating its distribution using a do operator, examining the effect of that intervention by querying marginals and resetting any interventions
```

Comment:
Style nit: this line is very long - wrap it to 2 lines?

Reply:
To add to this, if you like, you can format your python files using black for a more formatted look: https://black.readthedocs.io/en/stable/getting_started.html